Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp

```diff
@@ -0,0 +1,833 @@
+ * applicable because it states "Return CharCodeAt(string, index)", and CharCodeAt behaves as
+ * String.prototype.charCodeAt, which would return NaN given a negative index. I'm assuming the
+ * intended but underspecified behavior is to throw.
+ */
+
+DEFINE_BUILTIN_IMPLEMENTATION_I32(jsstring, charCodeAt, JSGlobalObject* globalObject, JSValue arg, int32_t index)
+{
+    auto scope = DECLARE_THROW_SCOPE(globalObject->vm());
+    if (!arg.isString()) [[unlikely]]
+        THROW_ILLEGAL_ARGUMENT_EXCEPTION;
+    auto* string = arg.toString(globalObject);
+    IF_EXCEPTION_RETHROW_AS_ILLEGAL_ARGUMENT(scope);
+    auto view = string->view(globalObject);
+    IF_EXCEPTION_RETHROW_AS_ILLEGAL_ARGUMENT(scope);
+    if (index < 0 || view->length() <= static_cast<unsigned>(index)) [[unlikely]]
+        THROW_ILLEGAL_ARGUMENT_EXCEPTION;
+    UChar result = view[index];
+    RELEASE_AND_RETURN(scope, result);
+}
+
+DEFINE_BUILTIN_ENTRY_I_RI(jsstring, charCodeAt)
+DEFINE_BUILTIN_JS_ENTRY_I_RI(jsstring, charCodeAt)
+DECLARE_BUILTIN_TRAMPOLINE(jsstring, charCodeAt);
+
+/**
+ * See https://webassembly.github.io/js-string-builtins/js-api/#js-string-codePointAt
+ *
+ * See the note on index validation in charCodeAt.
+ */
+
+DEFINE_BUILTIN_IMPLEMENTATION_I32(jsstring, codePointAt, JSGlobalObject* globalObject, JSValue arg, int32_t index)
+{
+    auto scope = DECLARE_THROW_SCOPE(globalObject->vm());
+    if (!arg.isString()) [[unlikely]]
+        THROW_ILLEGAL_ARGUMENT_EXCEPTION;
+    auto* string = arg.toString(globalObject);
+    IF_EXCEPTION_RETHROW_AS_ILLEGAL_ARGUMENT(scope);
+    auto view = string->view(globalObject);
+    IF_EXCEPTION_RETHROW_AS_ILLEGAL_ARGUMENT(scope);
+    unsigned length = view->length();
+    if (index < 0 || length <= static_cast<unsigned>(index)) [[unlikely]]
+        THROW_ILLEGAL_ARGUMENT_EXCEPTION;
+    int32_t result;
+    if (view->is8Bit())
+        result = view->span8()[index];
+    else {
+        char32_t character;
+        auto characters = view->span16();
+        unsigned i = static_cast<unsigned>(index);
+        U16_NEXT(characters, i, length, character);
+        result = static_cast<int32_t>(character);
+    }
+    RELEASE_AND_RETURN(scope, result);
+}
+
+DEFINE_BUILTIN_ENTRY_I_RI(jsstring, codePointAt)
+DEFINE_BUILTIN_JS_ENTRY_I_RI(jsstring, codePointAt)
+DECLARE_BUILTIN_TRAMPOLINE(jsstring, codePointAt);
+
+/**
+ * See https://webassembly.github.io/js-string-builtins/js-api/#js-string-length
+ *
+ * If the value is a string, return its length. Otherwise throw a RuntimeError.
+ */
+
+DEFINE_BUILTIN_IMPLEMENTATION_I32(jsstring, length, JSGlobalObject* globalObject, JSValue arg)
+{
+    auto scope = DECLARE_THROW_SCOPE(globalObject->vm());
+    if (!arg.isString()) [[unlikely]]
+        THROW_ILLEGAL_ARGUMENT_EXCEPTION;
+    unsigned length = asString(arg)->length();
+    if (length > static_cast<uint64_t>(std::numeric_limits<int32_t>::max())) [[unlikely]]
+        length = std::numeric_limits<int32_t>::max();
```

Comment:
This is always a no-op because `String::MaxLength == std::numeric_limits<int32_t>::max()`